controllers/control_plane_init_locker.go

```diff
@@ -18,52 +18,56 @@ package controllers
 
 import (
 	"context"
+	"encoding/json"
 	"fmt"
 
 	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
 	apicorev1 "k8s.io/api/core/v1"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	clusterv2 "sigs.k8s.io/cluster-api/api/v1alpha2"
+	"k8s.io/apimachinery/pkg/types"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha2"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
-// ControlPlaneInitLocker provides a locking mechanism for cluster initialization.
-type ControlPlaneInitLocker interface {
-	// Acquire returns true if it acquires the lock for the cluster.
-	Acquire(cluster *clusterv2.Cluster) bool
-}
+const lockInformationKey = "lock-information"
 
-// controlPlaneInitLocker uses a ConfigMap to synchronize cluster initialization.
-type controlPlaneInitLocker struct {
+// ControlPlaneInitLocker uses a ConfigMap to synchronize cluster initialization.
+type ControlPlaneInitLocker struct {
 	log    logr.Logger
 	client client.Client
-	ctx    context.Context
 }
 
-var _ ControlPlaneInitLocker = &controlPlaneInitLocker{}
-
-func newControlPlaneInitLocker(ctx context.Context, log logr.Logger, client client.Client) *controlPlaneInitLocker {
-	return &controlPlaneInitLocker{
-		ctx:    ctx,
+// NewControlPlaneInitLocker returns a lock that can be held by a control plane node before init.
+func NewControlPlaneInitLocker(log logr.Logger, client client.Client) *ControlPlaneInitLocker {
+	return &ControlPlaneInitLocker{
 		log:    log,
 		client: client,
 	}
 }
 
-func (l *controlPlaneInitLocker) Acquire(cluster *clusterv2.Cluster) bool {
-	configMapName := fmt.Sprintf("%s-controlplane", cluster.UID)
-	log := l.log.WithValues("namespace", cluster.Namespace, "cluster-name", cluster.Name, "configmap-name", configMapName)
+func lockName(clusterUID types.UID) string {
+	return fmt.Sprintf("%s-controlplane", clusterUID)
+}
+
+// Acquire allows a control plane node to be the first and only node to run kubeadm init
+func (c *ControlPlaneInitLocker) Acquire(ctx context.Context, cluster *clusterv1.Cluster, machine *clusterv1.Machine) bool {
+	configMapName := lockName(cluster.UID)
+	log := c.log.WithValues("namespace", cluster.Namespace, "cluster-name", cluster.Name, "configmap-name", configMapName)
 
-	exists, err := l.configMapExists(cluster.Namespace, configMapName)
+	exists, err := c.configMapExists(ctx, cluster.Namespace, configMapName)
 	if err != nil {
 		log.Error(err, "Error checking for control plane configmap lock existence")
 		return false
 	}
 	if exists {
 		return false
```

Comment:
In CAPA, we allow reentrant acquisition - see https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/2b93d46081179298cc62a15bac23a96b8c7695d9/pkg/cloud/aws/actuators/machine/control_plane_init_locker.go#L63-L71